### ISSUE TITLE
window and image sizes are incorrect on Mac retina and P2D

### ISSUE THREAD

**Issue by coreyackland** [processing/processing#5005]
Hi Guys, 

New to this GitHub and coding in general, especially with Processing.

Wondering why when I state my size(500,500); on Mac Retina, it seems the window size isn't actually 500x500. Its much larger.

I'm working on a resolution of 2880 x 1800. And when I set the size(2880,1800), the whole window goes completely offscreen as if its double the size of what I am asking.

I'm using the following:
size(2880,1800,P2D);
pixelDensity(2);

etc and still can get it to be correct for retina. 

The only way I could get the correct size is by reducing its size in half:

  image(BankOff,704,117,BankOff.width/2,BankOff.height/2);

I'm assuming there must be a more practical way than having to enter this in for every image I set in the window. I tried looking for this library (processing.core.PGraphicsRetina2D) and couldn't find it, nor download it through Processing. 

I'm searched the web and can't seem to find the correct solution. Hoping someone can help. Basically I'm laying out a load of images designed in photoshop for an interactive GUI. 

Thanks,
Corey

**Comment by benfry** [comment]
From your description it sounds like you're using a high-res display that's not an Apple-branded "retina" display, so the OS doesn't recognize it as such. Is that possible?

**Comment by coreyackland** [comment]
No it is apple, it's actually my Apple MacBook Pro Retina.

**Comment by benfry** [comment]
Huh… And this is with Processing 3.3?

**Comment by coreyackland** [comment]
Yes, correct. Processing 3.3

**Comment by JakubValtar** [comment]
Please try with 3.3.1, should be fixed in #5011.

**Comment by coreyackland** [comment]
Thanks for that, unfortunately upgrading to 3.3.1 doesn't seem to fix it.

With a simple test of :

void setup(){
size(500,500);
}

The processing window displayed (1000,1000), by taking a screenshot and finding the pixel dimensions.   So seems to double.

**Comment by JakubValtar** [comment]
I think this is how it is supposed to work. If you use 2x screen scaling your window will be 500 by 500 display units and 1000 by 1000 physical pixels. pixelWidth will be 500 and one pixel of sketch will correspond to 2x2 physical pixels on your screen.

With pixelDensity(2) the window should appear the same size as before, but the drawing surface will be 1000x1000 and pixelWidth will be 1000, pixels of sketch will match physical pixels of the screen.

Try to use `set(20, 20, #FF00FF);` and see how big the pixel is.

**Comment by coreyackland** [comment]
The pixel is small. I'm displaying and positioning GUI images within my app, they all seem to be double the size of what I designed them at. 

I can only get them to the correct size with this:

image(Panels,3,3,Panels.width/2,Panels.height/2);

**Comment by JakubValtar** [comment]
I think I understand now. I have two things to try, both of them should work. Run your sketch with `pixelDensity(2)` and

1) Use this to draw the GUI images:

```
set(x, y, image);
```

2) As a second alternative, run every GUI image through this once after loading it:

```
void makeImage2x(PImage img) {
  img.pixelDensity = 2;
  img.width /= 2;
  img.height /= 2;
}
```

**Comment by benfry** [comment]
@JakubValtar With `size(500, 500)` and `pixelDensity(2)`, the `pixelWidth` should be 1000, because `pixelWidth` and `pixelHeight` refer to the actual number of pixels (i.e. inside the `pixels[]` array). Is that a typo?

**Comment by JakubValtar** [comment]
@benfry What you wrote is correct. There was no typo but I updated my post to clarify.

**Comment by benfry** [comment]
Phew! I'll take my lack of reading comprehension over telling you the wrong thing to implement any day.

**Comment by benfry** [comment]
@coreyackland I'm realizing that you're missing how Apple does this stuff. Your laptop's screen size, as far as macOS is telling any applications, is 1440x900, not 2880x1800. The specifics of the number of pixels that are being used are completely hidden to apps, because you might be using any number of scaling options. 

Further, it looks like you probably just want `fullScreen(P2D)`. With `pixelDensity(displayDensity())` on your retina laptop, that'll give you a `width` and `height` of 1440 and 900, but the `pixelWidth` and `pixelHeight` will be 2880 and 1800, which is valid for any pixel-specific operations (`loadPixels()`, `get()`, `set()`, etc).

Bottom line, not a Processing bug; this is how macOS works.

**Comment by JakubValtar** [comment]
@benfry there is also another issue here: all loaded images are 1x and we don't have any way to specify that the image is 2x apart from doing it manually like I posted above (makeImage2x).

**Comment by benfry** [comment]
Yeah; some of that in this discussion: https://github.com/processing/processing/issues/3364
